src/test/ui/lint/lint-const-item-mutation.rs

```diff
@@ -0,0 +1,21 @@
+// check-pass
+
+struct MyStruct {
+    field: bool,
+    inner_array: [char; 1],
+}
+impl MyStruct {
+    fn use_mut(&mut self) {}
+}
+
+const ARRAY: [u8; 1] = [25];
+const MY_STRUCT: MyStruct = MyStruct { field: true, inner_array: ['a'] };
+
+fn main() {
+    ARRAY[0] = 5; //~ WARN attempting to modify
+    MY_STRUCT.field = false; //~ WARN attempting to modify
+    MY_STRUCT.inner_array[0] = 'b'; //~ WARN attempting to modify
+    MY_STRUCT.use_mut(); //~ WARN taking
+    &mut MY_STRUCT; //~ WARN taking
```

Comment:
I think your lint will also catch

```rust
let x = &mut MY_STRUCT;
```

which seems to me to be equivalent to `let x = &mut Default::default()` or similar things that (arguably rarely) are legitimite use cases.

Reply:
That's intentional - I think `let x = &mut MY_STRUCT` looks extremely misleading. If it there's ever the need for it in practice (maybe macro-generated code?), users can add `#[allow(const_item_mutation)]`.

However, this shouldn't become a hard error.